File: legacy/legacy-apis/Psi.Models.ClientConfigurationModels/Account/ScheduledTransfers.cs
using System;
using System.Collections.Generic;
using Psi.Business.ServiceContracts.RequestResponse.Transfers;
using Psi.Data.Models.Domain;
using Psi.Data.Models.Domain.Transfers;

namespace Psi.Data.Models.ClientConfigurationModels.Account
{
    public class ScheduledTransfers : SettingsBaseHelper
    {
        public ScheduledTransfers(ISettingsBase settingsBase) : base(settingsBase)
        {
        }

        [SettingKey("ScheduledTransfers.Enabled")]
        public bool Enabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.MinVersion")]
        public double MinVersion
        {
            get => GetDoubleValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.MonthsToCalculate")]
        public int MonthsToCalculate
        {
            get => GetIntValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.AdjustForWeekendsAndHolidays")]
        public bool AdjustForWeekendsAndHolidays => false;
        //{
        //TODO: Make this configurable when we have a plan to handle converting transfer exceptions from non-business days to business days.
        //get => GetBoolValue();
        //set => SetValue(value);
        //}

        [SettingKey("ScheduledTransfers.AddDayForWeekendOrHoliday")]
        public bool AddDayForWeekendOrHoliday
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.MaxTransferAttempts")]
        public int MaxTransferAttempts
        {
            get => GetIntValue();
            set => SetValue(value);
        }

        [SettingKey("X.App.HomeBanking.BillPayAccounts")]
        public string HomeBankingBillPayAccounts
        {
            get => GetValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.TransferAtLeastMinPaymentForLoans")]
        public bool TransferAtLeastMinPaymentForLoans
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.TransferAtLeastMinPaymentForCreditCards")]
        public bool TransferAtLeastMinPaymentForCreditCards
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ContinueIfMinPaymentCheckFails")]
        public bool ContinueIfMinPaymentCheckFails
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("X.App.HomeBanking.DefaultSuffixMask")]
        public string DefaultSuffixMask
        {
            get => GetValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.UseAmountDueInsteadOfPaymentForMinimumHelocPayment")]
        public bool ShouldUseAmountDueInsteadOfPaymentForMinimumHelocPayment
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.StartDateTimePickerHourOffset")]
        public int StartDateTimePickerHourOffset
        {
            get => GetIntValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.LastDayOfMonthSchedulingEnabled")]
        public bool LastDayOfMonthSchedulingEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.CoreErrorCodeToTransferResultMappings")]
        public Dictionary<string, ScheduledTransferResult> CoreErrorCodeToTransferResultMappings
        {
            get
            {
                var value = GetJsonValueOrNull<Dictionary<string, ScheduledTransferResult>>();

                return value ?? new Dictionary<string, ScheduledTransferResult>();
            }
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.CoreErrorCodeToTransferResultMappingEnabled")]
        public bool CoreErrorCodeToTransferResultMappingEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.UseScheduledRunTimeAsEffectiveDate.Enabled")]
        public bool UseScheduledRunTimeAsEffectiveDateEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [Obsolete("This isn't used right now. A future project may be completed that uses this setting.")]
        [SettingKey("ScheduledTransfers.CheckCoreForDuplicateTransactions.Enabled")]
        public bool CheckCoreForDuplicateTransactionsEnabled
        {
            get => false;
            set => SetValue(value);
        }

        [Obsolete("This isn't used right now. A future project may be completed that uses this setting.")]
        [SettingKey("ScheduledTransfers.CheckCoreForDuplicateTransactions.ShouldSendSecureMessageWhenDuplicateIsDetected")]
        public bool ShouldSendSecureMessageWhenDuplicateIsDetected
        {
            get => false;
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ReadOnly.Enabled")]
        public bool ReadOnlyEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ValidateProcessingIndicatorsEnabled")]
        public bool ValidateProcessingIndicatorsEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ShouldSendSecureMessageWhenManualReviewIsRequired")]
        public bool ShouldSendSecureMessageWhenManualReviewIsRequired
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ManualReviewRequiredSecureMessageCategory")]
        public string ManualReviewRequiredSecureMessageCategory
        {
            get => GetValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ManualReviewRequiredSecureMessageSenderMembershipUserId")]
        public int ManualReviewRequiredSecureMessageSenderMembershipUserId
        {
            get => GetIntValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ShouldStoreIdentifierForCreditCards")]
        public bool ShouldStoreIdentifierForCreditCards
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.InternalTransfers.RecurrenceTypes")]
        public List<RecurrenceType> InternalTransfersRecurrenceTypes
        {
            get
            {
                var recurrenceTypes = GetListValue() ?? new List<string>();
                var result = new List<RecurrenceType>();

                foreach (var recurrenceType in recurrenceTypes)
                {
                    if (Enum.TryParse(recurrenceType, true, out RecurrenceType type))
                    {
                        result.Add(type);
                    }
                }

                return result;
            }
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ExternalTransfers.RecurrenceTypes")]
        public List<RecurrenceType> ExternalTransfersRecurrenceTypes
        {
            get
            {
                var recurrenceTypes = GetListValue() ?? new List<string>();
                var result = new List<RecurrenceType>();

                foreach (var recurrenceType in recurrenceTypes)
                {
                    if (Enum.TryParse(recurrenceType, true, out RecurrenceType type))
                    {
                        result.Add(type);
                    }
                }

                return result;
            }
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.CreditCardTransfers.RecurrenceTypes")]
        public List<RecurrenceType> CreditCardTransfersRecurrenceTypes
        {
            get
            {
                var recurrenceTypes = GetListValue() ?? new List<string>();
                var result = new List<RecurrenceType>();

                foreach (var recurrenceType in recurrenceTypes)
                {
                    if (Enum.TryParse(recurrenceType, true, out RecurrenceType type))
                    {
                        result.Add(type);
                    }
                }

                return result;
            }
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.EndOptions")]
        public List<ScheduledTransferEndOption> EndOptions
        {
            get
            {
                var endOptions = GetListValue() ?? new List<string>();
                var result = new List<ScheduledTransferEndOption>();

                foreach (var endOption in endOptions)
                {
                    if (Enum.TryParse(endOption, true, out ScheduledTransferEndOption option))
                    {
                        result.Add(option);
                    }
                }

                return result;
            }
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.OnHost.Enabled")]
        public bool OnHostEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.OnHost.MinVersion")]
        public double OnHostMinVersion
        {
            get => GetDoubleValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.AutomaticLoanPayments.Enabled")]
        public bool AutomaticLoanPaymentsEnabled
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.AutomaticLoanPayments.LoanCategories")]
        public List<string> AutomaticLoanPaymentsLoanCategories
        {
            get => GetListValue() ?? new List<string>();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.ShouldUseNewUserInterface")]
        public bool ShouldUseNewUserInterface
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.CanSkipTransferOccurrence")]
        public bool CanSkipTransferOccurrence
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("ScheduledTransfers.CanEditTransferAmount")]
        public bool CanEditTransferAmount
        {
            get => GetBoolValue();
            set => SetValue(value);
        }
    }
}

File: workspace/legacy-analyzer/output/infrastructure/models/Account/ScheduledTransfers.ts
// Generated imports
import { CoreErrorCodeToTransferResultMappings } from '../CoreErrorCodeToTransferResultMappings';
import { InternalTransfersRecurrenceTypes } from '../InternalTransfersRecurrenceTypes';
import { ExternalTransfersRecurrenceTypes } from '../ExternalTransfersRecurrenceTypes';
import { CreditCardTransfersRecurrenceTypes } from '../CreditCardTransfersRecurrenceTypes';
import { EndOptions } from '../EndOptions';
import { AutomaticLoanPaymentsLoanCategories } from '../AutomaticLoanPaymentsLoanCategories';

export interface ScheduledTransfers {
    /** @settingKey ScheduledTransfers.Enabled */
    enabled: boolean;
    /** @settingKey ScheduledTransfers.MinVersion */
    minVersion: number;
    /** @settingKey ScheduledTransfers.MonthsToCalculate */
    monthsToCalculate: number;
    /** @settingKey ScheduledTransfers.AdjustForWeekendsAndHolidays */
    adjustForWeekendsAndHolidays: boolean;
    /** @settingKey ScheduledTransfers.AddDayForWeekendOrHoliday */
    /**
     * /// <summary>
     * /// //{
     * /// //TODO: Make this configurable when we have a plan to handle converting transfer exceptions from non-business days to business days.
     * /// //get => GetBoolValue();
     * /// //set => SetValue(value);
     * /// //}
     * /// </summary>
     */
    addDayForWeekendOrHoliday: boolean;
    /** @settingKey ScheduledTransfers.MaxTransferAttempts */
    maxTransferAttempts: number;
    /** @settingKey X.App.HomeBanking.BillPayAccounts */
    homeBankingBillPayAccounts: string;
    /** @settingKey ScheduledTransfers.TransferAtLeastMinPaymentForLoans */
    transferAtLeastMinPaymentForLoans: boolean;
    /** @settingKey ScheduledTransfers.TransferAtLeastMinPaymentForCreditCards */
    transferAtLeastMinPaymentForCreditCards: boolean;
    /** @settingKey ScheduledTransfers.ContinueIfMinPaymentCheckFails */
    continueIfMinPaymentCheckFails: boolean;
    /** @settingKey X.App.HomeBanking.DefaultSuffixMask */
    defaultSuffixMask: string;
    /** @settingKey ScheduledTransfers.UseAmountDueInsteadOfPaymentForMinimumHelocPayment */
    shouldUseAmountDueInsteadOfPaymentForMinimumHelocPayment: boolean;
    /** @settingKey ScheduledTransfers.StartDateTimePickerHourOffset */
    startDateTimePickerHourOffset: number;
    /** @settingKey ScheduledTransfers.LastDayOfMonthSchedulingEnabled */
    lastDayOfMonthSchedulingEnabled: boolean;
    /** @settingKey ScheduledTransfers.CoreErrorCodeToTransferResultMappings */
    dictionary: CoreErrorCodeToTransferResultMappings;
    /** @settingKey ScheduledTransfers.CoreErrorCodeToTransferResultMappingEnabled */
    coreErrorCodeToTransferResultMappingEnabled: boolean;
    /** @settingKey ScheduledTransfers.UseScheduledRunTimeAsEffectiveDate.Enabled */
    useScheduledRunTimeAsEffectiveDateEnabled: boolean;
    /** @settingKey ScheduledTransfers.CheckCoreForDuplicateTransactions.Enabled */
    checkCoreForDuplicateTransactionsEnabled: boolean;
    /** @settingKey ScheduledTransfers.CheckCoreForDuplicateTransactions.ShouldSendSecureMessageWhenDuplicateIsDetected */
    shouldSendSecureMessageWhenDuplicateIsDetected: boolean;
    /** @settingKey ScheduledTransfers.ReadOnly.Enabled */
    readOnlyEnabled: boolean;
    /** @settingKey ScheduledTransfers.ValidateProcessingIndicatorsEnabled */
    validateProcessingIndicatorsEnabled: boolean;
    /** @settingKey ScheduledTransfers.ShouldSendSecureMessageWhenManualReviewIsRequired */
    shouldSendSecureMessageWhenManualReviewIsRequired: boolean;
    /** @settingKey ScheduledTransfers.ManualReviewRequiredSecureMessageCategory */
    manualReviewRequiredSecureMessageCategory: string;
    /** @settingKey ScheduledTransfers.ManualReviewRequiredSecureMessageSenderMembershipUserId */
    manualReviewRequiredSecureMessageSenderMembershipUserId: number;
    /** @settingKey ScheduledTransfers.ShouldStoreIdentifierForCreditCards */
    shouldStoreIdentifierForCreditCards: boolean;
    /** @settingKey ScheduledTransfers.InternalTransfers.RecurrenceTypes */
    list: InternalTransfersRecurrenceTypes;
    /** @settingKey ScheduledTransfers.ExternalTransfers.RecurrenceTypes */
    list: ExternalTransfersRecurrenceTypes;
    /** @settingKey ScheduledTransfers.CreditCardTransfers.RecurrenceTypes */
    list: CreditCardTransfersRecurrenceTypes;
    /** @settingKey ScheduledTransfers.EndOptions */
    list: EndOptions;
    /** @settingKey ScheduledTransfers.OnHost.Enabled */
    onHostEnabled: boolean;
    /** @settingKey ScheduledTransfers.OnHost.MinVersion */
    onHostMinVersion: number;
    /** @settingKey ScheduledTransfers.AutomaticLoanPayments.Enabled */
    automaticLoanPaymentsEnabled: boolean;
    /** @settingKey ScheduledTransfers.AutomaticLoanPayments.LoanCategories */
    list: AutomaticLoanPaymentsLoanCategories;
    /** @settingKey ScheduledTransfers.ShouldUseNewUserInterface */
    shouldUseNewUserInterface: boolean;
    /** @settingKey ScheduledTransfers.CanSkipTransferOccurrence */
    canSkipTransferOccurrence: boolean;
    /** @settingKey ScheduledTransfers.CanEditTransferAmount */
    canEditTransferAmount: boolean;
}
